yew-ui/src/model/client/video_call_client.rs

```diff
@@ -0,0 +1,318 @@
+use super::super::connection::{ConnectOptions, Connection};
+use super::super::decode::PeerDecodeManager;
+use crate::crypto::aes::Aes128State;
+use crate::crypto::rsa::RsaWrapper;
+use anyhow::{anyhow, Result};
+use log::{debug, error, info};
+use protobuf::Message;
+use rsa::pkcs8::{DecodePublicKey, EncodePublicKey};
+use rsa::RsaPublicKey;
+use std::cell::RefCell;
+use std::rc::{Rc, Weak};
+use std::sync::Arc;
+use types::protos::aes_packet::AesPacket;
+use types::protos::media_packet::media_packet::MediaType;
+use types::protos::packet_wrapper::packet_wrapper::PacketType;
+use types::protos::packet_wrapper::PacketWrapper;
+use types::protos::rsa_packet::RsaPacket;
+use yew::prelude::Callback;
+#[derive(Clone, Debug, PartialEq)]
+pub struct VideoCallClientOptions {
+    pub enable_e2ee: bool,
+    pub enable_webtransport: bool,
+    pub on_peer_added: Callback<String>,
+    pub on_peer_first_frame: Callback<(String, MediaType)>,
+    pub get_peer_video_canvas_id: Callback<String, String>,
+    pub get_peer_screen_canvas_id: Callback<String, String>,
+    pub userid: String,
+    pub websocket_url: String,
+    pub webtransport_url: String,
+    pub on_connected: Callback<()>,
+    pub on_connection_lost: Callback<()>,
+}
+
+#[derive(Debug)]
+struct InnerOptions {
+    enable_e2ee: bool,
+    userid: String,
+}
+
+#[derive(Debug)]
+struct Inner {
+    options: InnerOptions,
+    connection: Option<Connection>,
+    aes: Arc<Aes128State>,
+    rsa: Arc<RsaWrapper>,
+    peer_decode_manager: PeerDecodeManager,
+}
+
+#[derive(Clone, Debug)]
+pub struct VideoCallClient {
+    options: VideoCallClientOptions,
+    inner: Rc<RefCell<Inner>>,
+}
+
+impl PartialEq for VideoCallClient {
+    fn eq(&self, other: &Self) -> bool {
+        Rc::ptr_eq(&self.inner, &other.inner) && self.options == other.options
+    }
+}
+
+impl VideoCallClient {
+    pub fn new(options: VideoCallClientOptions) -> Self {
+        let inner = Rc::new(RefCell::new(Inner {
```

Comment:
Hey @ronen thanks for your huge contributions like always! 

I understand why we are using the RefCell here, so that we can get mutable access to the Inner object, please consider wrapping Inner with RwLock so that we do not miss the compile-time checks.

Reply:
I didn't know about RwLock, but I did try Mutex -- but unfortunately either of those requires that the wrapped object satisfies `Send`.   And unfortunately somewhere deep down in the connection task objects there are members that don't satisfy Send.   I imagine there's gotta be a way to work around that, but nothing came to mind immediately and I decided not to go too far down that rabbit hole.

For RefCell I only use try_borrow and try_borrow_mut and I check the results, so it won't panic.  But if the try_borrow fails it doesn't block, it simply skips the operation it was supposed to do... so still not necessarily completely robust: it may drop an operation if there is another operation in progress in another JS "thread" -- so ultimately I think using the sync locks would be better if somebody.   [There was one case where I used `borrow()` which I think would never have failed, I just pushed another commit to obviate it]

Reply:
Give a chance to review this, I think we can replace RefCell if we change the design pattern a bit, this is a common issue in rust

Reply:
> Give a chance to review this, I think we can replace RefCell if we change the design pattern a bit, this is a common issue in rust

That'd be great!

I've been deliberately making incremental changes to the original code.  But that didn't necessarily walk me down the best path for the resulting design of all these modules.   A re-think of the design as a whole might well do better.